src/main/java/com/hubspot/jinjava/util/EagerContextWatcher.java

```diff
@@ -0,0 +1,468 @@
+    return speculativeBindings;
+  }
+
+  private static void cacheRevertibleObject(
+    JinjavaInterpreter interpreter,
+    Map<String, Object> initiallyResolvedHashes,
+    Map<String, String> initiallyResolvedAsStrings,
+    Entry<String, Object> entry
+  ) {
+    RevertibleObject revertibleObject = interpreter
+      .getRevertibleObjects()
+      .get(entry.getKey());
+    Object hashCode = initiallyResolvedHashes.get(entry.getKey());
+    try {
+      if (revertibleObject == null || !hashCode.equals(revertibleObject.getHashCode())) {
+        revertibleObject =
+          new RevertibleObject(
+            hashCode,
+            PyishObjectMapper.getAsPyishStringOrThrow(entry.getValue())
+          );
+        interpreter.getRevertibleObjects().put(entry.getKey(), revertibleObject);
+      }
+      revertibleObject
+        .getPyishString()
+        .ifPresent(
+          pyishString -> initiallyResolvedAsStrings.put(entry.getKey(), pyishString)
+        );
+    } catch (Exception e) {
+      interpreter
+        .getRevertibleObjects()
+        .put(entry.getKey(), new RevertibleObject(hashCode));
+    }
+  }
+
+  private static Object getOriginalValue(
+    JinjavaInterpreter interpreter,
+    EagerChildContextConfig eagerChildContextConfig,
+    Map<String, Object> initiallyResolvedHashes,
+    Map<String, String> initiallyResolvedAsStrings,
+    Entry<String, Object> e
+  ) {
+    if (eagerChildContextConfig.takeNewValue) {
+      if (e.getValue() instanceof DeferredValue) {
+        return ((DeferredValue) e.getValue()).getOriginalValue();
+      }
+      return e.getValue();
+    }
+
+    if (
+      e.getValue() instanceof DeferredValue &&
+      initiallyResolvedHashes
+        .get(e.getKey())
+        .equals(getObjectOrHashCode(((DeferredValue) e.getValue()).getOriginalValue()))
+    ) {
+      return ((DeferredValue) e.getValue()).getOriginalValue();
+    }
+
+    // This is necessary if a state-changing function, such as .update()
+    // or .append() is run against a variable in the context.
+    // It will revert the effects when takeNewValue is false.
+    if (initiallyResolvedAsStrings.containsKey(e.getKey())) {
+      // convert to new list or map
+      try {
+        return interpreter.resolveELExpression(
+          initiallyResolvedAsStrings.get(e.getKey()),
+          interpreter.getLineNumber()
+        );
+      } catch (DeferredValueException ignored) {}
+    }
+
+    // Previous value could not be mapped to a string
+    throw new CannotReconstructValueException(e.getKey());
+  }
+
+  private static Object getObjectOrHashCode(Object o) {
+    if (o instanceof LazyExpression) {
+      o = ((LazyExpression) o).get();
+    }
+    if (o instanceof PyList && !((PyList) o).toList().contains(o)) {
+      return o.hashCode();
+    }
+    if (o instanceof PyMap && !((PyMap) o).toMap().containsValue(o)) {
+      return o.hashCode() + ((PyMap) o).keySet().hashCode();
```

Comment:
It isn't because the implementation for a HashMap node's hashcode is:
```
return Objects.hashCode(key) ^ Objects.hashCode(value)
```
So if the key and value are the same, then the hash code will be 0. This would result in a map of `{'a': 'a'}` having the same hashcode as `{'b': 'b'}` or `{'a': 'a', 'b': 'b', 'c': 'c'}`, etc. They all have a hash code of 0.

Reply:
Good. There are hash collisions. We are not relying on the hashcode for equality, right?

Reply:
HashMap's `equals()` is not broken so we don't have to worry about that:
```
(Objects.equals(key, e.getKey()) && Objects.equals(value, e.getValue()))
```